libp2p/protocols/kademlia/kademlia.nim

```diff
@@ -17,14 +24,142 @@ type KadDHT* = ref object of LPProtocol
   rtable*: RoutingTable
   maintenanceLoop: Future[void]
 
+proc sendFindNode(
+    kad: KadDHT, peerId: PeerId, addrs: seq[MultiAddress], targetId: Key
+): Future[Message] {.
+    async: (raises: [CancelledError, DialFailedError, ValueError, LPStreamError])
+.} =
+  var conn: Connection
+  if addrs.len == 0:
+    conn = await kad.switch.dial(peerId, KadCodec)
+  else:
+    conn = await kad.switch.dial(peerId, addrs, KadCodec)
+
+  defer:
+    await conn.close()
+
+  let msg = Message(msgType: MessageType.findNode, key: some(targetId.getBytes()))
+
+  await conn.writeLp(msg.encode().buffer)
+
+  let reply = Message.decode(await conn.readLp(4096)).tryGet()
+
+  if reply.msgType != MessageType.findNode:
+    raise newException(ValueError, "unexpected message type in reply: " & $reply)
+
+  return reply
+
+proc waitRepliesOrTimeouts(
+    pendingFutures: Table[PeerId, Future[Message]]
+): Future[(seq[Message], seq[PeerId])] {.async: (raises: [CancelledError]).} =
+  await allFutures(toSeq(pendingFutures.values))
+
+  var receivedReplies: seq[Message] = @[]
+  var failedPeers: seq[PeerId] = @[]
+
+  for (peerId, replyFut) in pendingFutures.pairs:
+    try:
+      let reply = await replyFut
+      receivedReplies.add(reply)
+    except CatchableError:
+      failedPeers.add(peerId)
+      error "could not send find_node to peer", peerId, err = getCurrentExceptionMsg()
+
+  return (receivedReplies, failedPeers)
+
+proc findNode*(
+    kad: KadDHT, targetId: Key
+): Future[seq[PeerId]] {.async: (raises: [CancelledError]).} =
+  #debug "findNode", target = target
+  # TODO: should it return a single peer instead? read spec
+
+  var initialPeers = kad.rtable.findClosestPeers(targetId, DefaultReplic)
+  var state = LookupState.init(targetId, initialPeers)
+  var addrTable: Table[PeerId, seq[MultiAddress]] =
+    initTable[PeerId, seq[MultiAddress]]()
+
+  while not state.done:
+    let toQuery = state.selectAlphaPeers()
+
+    var pendingFutures = initTable[PeerId, Future[Message]]()
```

Comment:
Nitity nit: coud we maybe have a `var` that we initialize once and clear in every iteration instead? There's maybe not much gain though